File: benchmark-framework-source/src/main/java/com/benchmark/metrics/MetricsCollector.java
package com.benchmark.metrics;

import java.lang.management.ManagementFactory;
import java.lang.management.MemoryMXBean;
import java.lang.management.MemoryUsage;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import com.sun.management.OperatingSystemMXBean;

/**
 * Runs a background sampling thread that collects CPU and memory snapshots
 * at a configurable interval.
 *
 * <p>Usage:
 * <pre>{@code
 *   MetricsCollector collector = new MetricsCollector(200);
 *   collector.start();
 *   // ... run workload ...
 *   List<MetricsSnapshot> samples = collector.stop();
 * }</pre>
 * </p>
 */
public final class MetricsCollector {

    private final int                       sampleIntervalMs;
    private final OperatingSystemMXBean     osMXBean;
    private final MemoryMXBean              memMXBean;
    private final List<MetricsSnapshot>     snapshots = new ArrayList<>();
    private final ScheduledExecutorService  scheduler;
    private ScheduledFuture<?>              future;

    public MetricsCollector(int sampleIntervalMs) {
        this.sampleIntervalMs = sampleIntervalMs;
        this.osMXBean  = (OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();
        this.memMXBean = ManagementFactory.getMemoryMXBean();
        this.scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "metrics-collector");
            t.setDaemon(true);
            return t;
        });
    }

    /** Begin periodic sampling. */
    public synchronized void start() {
        snapshots.clear();
        future = scheduler.scheduleAtFixedRate(
                this::sample,
                0,
                sampleIntervalMs,
                TimeUnit.MILLISECONDS
        );
    }

    /** Stop sampling and return all collected snapshots (immutable). */
    public synchronized List<MetricsSnapshot> stop() {
        if (future != null) {
            future.cancel(false);
            future = null;
        }
        // One final sample
        sample();
        return Collections.unmodifiableList(new ArrayList<>(snapshots));
    }

    /** Shut down the background scheduler. Call when done with this collector. */
    public void shutdown() {
        scheduler.shutdownNow();
    }

    // -------------------------------------------------------------------------

    private synchronized void sample() {
        double cpuUsage = osMXBean.getProcessCpuLoad() * 100.0;
        if (cpuUsage < 0) cpuUsage = 0; // -1 means not yet available

        MemoryUsage heap    = memMXBean.getHeapMemoryUsage();
        MemoryUsage nonHeap = memMXBean.getNonHeapMemoryUsage();

        snapshots.add(new MetricsSnapshot(
                System.currentTimeMillis(),
                cpuUsage,
                heap.getUsed(),
                heap.getMax(),
                nonHeap.getUsed()
        ));
    }

    // -------------------------------------------------------------------------
    // Aggregate helpers used by BenchmarkResult

    public static double averageCpu(List<MetricsSnapshot> samples) {
        return samples.stream()
                .mapToDouble(MetricsSnapshot::getCpuUsagePercent)
                .average().orElse(0);
    }

    public static double peakCpu(List<MetricsSnapshot> samples) {
        return samples.stream()
                .mapToDouble(MetricsSnapshot::getCpuUsagePercent)
                .max().orElse(0);
    }

    public static double averageMemoryMB(List<MetricsSnapshot> samples) {
        return samples.stream()
                .mapToDouble(MetricsSnapshot::getTotalMemoryMB)
                .average().orElse(0);
    }

    public static double peakMemoryMB(List<MetricsSnapshot> samples) {
        return samples.stream()
                .mapToDouble(MetricsSnapshot::getTotalMemoryMB)
                .max().orElse(0);
    }
}

File: python/benchmark/metrics/metrics_collector.py
from __future__ import annotations

import threading
import time
from typing import List

import psutil

from benchmark.metrics.metrics_snapshot import MetricsSnapshot


class MetricsCollector:
    """
    Background sampling of CPU and RSS memory at a configurable interval.
    Maps RSS into heap_used_bytes; non_heap is 0 (parity with total MB reporting).
    """

    def __init__(self, sample_interval_ms: int) -> None:
        self._sample_interval_ms = sample_interval_ms
        self._snapshots: List[MetricsSnapshot] = []
        self._lock = threading.Lock()
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None
        self._process = psutil.Process()

    def start(self) -> None:
        with self._lock:
            self._snapshots.clear()
        self._stop.clear()
        self._process.cpu_percent(interval=None)
        self._thread = threading.Thread(target=self._run, name="metrics-collector", daemon=True)
        self._thread.start()

    def stop(self) -> tuple[MetricsSnapshot, ...]:
        self._stop.set()
        if self._thread is not None:
            self._thread.join(timeout=self._sample_interval_ms / 1000.0 * 5 + 1.0)
            self._thread = None
        self._sample()
        with self._lock:
            return tuple(self._snapshots)

    def shutdown(self) -> None:
        self._stop.set()

    def _run(self) -> None:
        interval_s = self._sample_interval_ms / 1000.0
        while not self._stop.is_set():
            self._sample()
            if self._stop.wait(timeout=interval_s):
                break

    def _sample(self) -> None:
        cpu = self._process.cpu_percent(interval=None)
        if cpu < 0:
            cpu = 0.0
        rss = int(self._process.memory_info().rss)
        snap = MetricsSnapshot(
            timestamp_ms=int(time.time() * 1000),
            cpu_usage_percent=cpu,
            heap_used_bytes=rss,
            heap_max_bytes=0,
            non_heap_used_bytes=0,
        )
        with self._lock:
            self._snapshots.append(snap)

    @staticmethod
    def average_cpu(samples: tuple[MetricsSnapshot, ...]) -> float:
        if not samples:
            return 0.0
        return sum(s.cpu_usage_percent for s in samples) / len(samples)

    @staticmethod
    def peak_cpu(samples: tuple[MetricsSnapshot, ...]) -> float:
        if not samples:
            return 0.0
        return max(s.cpu_usage_percent for s in samples)

    @staticmethod
    def average_memory_mb(samples: tuple[MetricsSnapshot, ...]) -> float:
        if not samples:
            return 0.0
        return sum(s.get_total_memory_mb() for s in samples) / len(samples)

    @staticmethod
    def peak_memory_mb(samples: tuple[MetricsSnapshot, ...]) -> float:
        if not samples:
            return 0.0
        return max(s.get_total_memory_mb() for s in samples)
